### ISSUE TITLE
tiles cache limitations?

### ISSUE THREAD

**Issue by jordyno** [google/marzipano#110]
Hello guys,
are there any tile cache limitations? On my implementation and also in your example `http://www.marzipano.net/demos/sample-tour/index.html` , it looks like that tiles get thrown away after certain amount is loaded. For example, on the link above, when you zoom in almost all the way in, then slowly look around 360 degrees while waiting for all the hires tiles to load, when you quickly turn around 180 degrees the low-res tiles are shown and I have to wait until the hires ones finish loading again. Or is there another reason for this?
Thanks

**Comment by tjgq** [comment]
The caching logic is in the `TextureStore` class. In a nutshell:

- Visible tiles are always kept in the cache.
- Pinned tiles are always kept in the cache (usually the preview level is pinned so that there is always a fall back of last resort).
- There is an LRU cache of tiles that used to be visible, but no longer are. By default, the cache holds 32 tiles, but this can be changed by passing the `previouslyVisibleCacheSize` option into the `TextureStore`.

For gigapixel panoramas, this means that if you zoom in too far (burning through the entire 32-tile LRU cache) and then move outside of that area, you will see the first level (because it's pinned) until the larger levels load again.

The main challenge with increasing the cache size too much is that memory limitations vary considerably for different devices, and there's no way to determine how much memory is available. That said, 32 might be a bit too conservative even for today's low-end hardware.

**Comment by jordyno** [comment]
In my implementation, for some layers, I use transparent overalys with empty tiles being ignored, e.g. empty.png tile is returned. Does such repeated tile use the limit set by previouslyVisibleCacheSize? I mean, if I have 32 tiles loaded is it the case that half of them can be just empty.png taking up nearly no memory at all? 
If that is true, would probably make sense to set previouslyVisibleCacheSize as per layer, so that the limit is higher on such layers where more empty tiles are expected.
Just to make sure, if I have empty.png tile, which is 1096 bytes only, is it exactly that amount what it takes up in memory, or is it more as it gets placed onto canvas and multiplied by how many times it is used?

**Comment by tjgq** [comment]
Each TextureStore instance gets its own LRU cache, and there should be a separate TextureStore instance per Layer (your setup code should instantiate a fresh one to pass to the Layer constructor).

The TextureStore caches WebGL textures, which take up memory proportional to the tile size, not the file size (and even more if mipmapping is enabled, though it's disabled by default). The cache key is the tile, not the file the texture came from. In your case this is unfortunate, since every one of your identical transparent tiles will be stored separately.

**Comment by jordyno** [comment]
So, if I understand you right, replacing empty tiles with one identical tile only saves bandwidth, but not processing or memory resources. Would there be some clever way of making TextureStore ignore specific tiles? I already do have an array of empty tiles per each layer. Even if I was able to ignore the entire cubeface only, that would help too, since the sky needs no overlays in my scenario.

**Comment by tjgq** [comment]
Unfortunately, there's no way to "ignore" tiles in the current implementation. We could maybe add a special tile attribute that says "use a default transparent texture", but I haven't thought through all the implications.

Here's a list of things that I would try to mitigate the issue:

- Increase the LRU size. I would test on lower-end devices to make sure there are no memory issues.
- Make your transparent images 1x1 . The image and tile size don't have to match; textures are created at the image size, and scaled up during rendering if necessary. This won't prevent multiple textures from being created, but at least they will be a lot smaller.
- Make sure that your browser doesn't fetch the transparent PNG more than once. If you see multiple requests for it, this could mean your web server is configured incorrectly and is preventing the browser from caching the file.
- Increase the tile size to reduce pressure on the LRU cache.

If any of these steps make the situation better, let us know. It's valuable information to guide future optimizations.

**Comment by jordyno** [comment]
What I don't understand is what happens on the WebGL side if a specific tile is not found. From my perspective, I am getting just GET errors, but the viewer continues working normally.
Quick test showed that Firefox took approx. equal amount of system memory when fed empty.png tile and when failed tile GET requests were unhandled.

In an ideal world, I would rather keep my graphic items separately on each layer ( = less used tiles per level, but many - 10~20 layers per scene) rather than trying to fit as many graphic items on one layer ( = more tiles per layer, less layers), because then I could easily modify say, "paths" using layer effects. But since empty tiles also consume LRU, I guess having more than 5 layers on 5 scenes would be an overkill.

The main problem in my situation is, that LruSet as expected, does not consider type of tile being added. So it seems that non-empty tiles get replaced by empty tiles only because they are "older". If we could somehow prioritize tiles inside LruSet, or check if their size is say, 1x1 px (for empty tiles) and add/remove tiles acordingly, it would be a solution. Pinning non-empty tiles might also work, but I guess it would be very memory-intensive unless it is possible to pin only specific tiles per specific level.

Will continue testing as you suggested and keep you updated. In the meantime:

- I am already using 1024px cubefaces for PNG layers. Some of the tiles reach up to 150kb, which is already quite a lot, so I am thinking about going back to 512px.
- Monitoring network showed that the browser is requesting the empty.png tile only once (until LRU is full), so that is good.
- Increasing LRU size certainly helps, but when looking around scene, the empty tiles will eventually accelerate replacing of the the filled tiles and they have to be loaded again.

Mainly, what I am looking into is how to change LRU behavior so that empty tiles do not get added, if that is possible. Maybe some changes to LruSet.prototype.add could do the trick?

**Comment by jordyno** [comment]
One more idea: How is a pinned tile related to LRU? Does the previouslyVisibleCacheSize number include pinned tiles? If I have a first level pinned (6 tiles), does that give me 32-6=24 default available tile spots for LRU cache? Or is this value independent of pinned tiles?
Simplest approach for me would be modifying Layer.js and creating a pinTiles prototype to which I would pass an Array of tiles I would like to pin. This way I could pin level 1 or 2 without worrying about memory limitations too much, what do you think?
Still, the best solution could be modifying LRU behavior so that empty tiles do not get added into the stack as they do not need loading anyway, if I am not mistaken. But this seems to be beyond my current programming skills.

**Comment by tjgq** [comment]
If the file cannot be fetched, the tile is not rendered (ordinarily, the stage would fall back into parent or children tiles if they're available, but in this case they won't be). After a timeout, the TextureStore will try to fetch the file again. So apart from the unnecessary network traffic, this does have the effect that you want (transparent tiles take no space in the LRU).

Assuming the devices you want to support have enough memory, you could try pinning another level or two (use `Layer#pinLevel`). Pinned tiles don't count towards the LRU limit.

I think a proper fix would involve either (1) teaching the TextureStore not to load identical textures twice, or (2) teaching the TextureStore and Stage to recognize an asset as transparent and exclude it from rendering. Either way would ensure that a transparent texture is never cached more than once.

**Comment by jordyno** [comment]
Wouldn't it be possible to make a condition not to load the tile somewhere [here](https://github.com/google/marzipano/blob/3194c4f6dea9aa3929319d68f4f65ce4d4318a49/src/TextureStore.js#L101) in case the source filepath is an empty string? In such case I would just send to ImageUrlSource empty string instead of link to empty tile png?
I guess this could be a win-win situation where I could split all the graphic elements into many layers without too big of an impact on memory and processing resources?

**Comment by tjgq** [comment]
Yes, that would be option (2) above. It's a bit more involved than just skipping texture loading; the TextureStore should also not repeatedly try to reload it on every frame, and the Stage should also not try to find a parent or child fallback.

The thing that bothers me about adding that logic is that it targets an extremely narrow use case. For example, what if the pano background is tinted red (e.g. `rgba(255, 0, 0, 0.1)`) instead of fully transparent? It would still be nice not to load dozens of identical red-tinted textures, but special-casing transparent ones won't help with that. Option (1) would be a better long-term solution.

**Comment by jordyno** [comment]
You are absolutely right. Option (1) makes more sense also in different situations than mine. 
Still, option (2) sounds more memory and CPU resources conservative, especially with more layers on stage, what do you think? The best would be probably a combination of both, but I have no idea how to implement it, sounds difficult. Could you please give me some lead?

**Comment by tjgq** [comment]
For option (2), the simplest thing to do is add a `Source#hasAsset(tile)` method that returns false if the tile is transparent. Then, in `TextureStore#_loadTile`, return without doing anything if that's the case. This can be optimized further, but at least it prevents transparent textures from being created and filling up the LRU.

**Comment by jordyno** [comment]
Many thanks Tiago! Will try it out and test how it affects the performance and then post it here. I am currently running with 1 panorama + 3 transparent layers and the viewer performance difference with single layers (panorama only) is noticable. But now as the alpha blending is improved, I might go with 1 level lower per layer, as the tiles look okay even at bit lower resolution.

**Comment by jordyno** [comment]
it really makes sense to have this feature, the performance gain is quite noticable!

**Comment by jordyno** [comment]
Tiago, what was your idea of improving this further, do you think there can be more CPU/mem resources conservation?
